### DISCUSSION TITLE
Term translation in the Java bindings

### DISCUSSION THREAD

**Discussion by baierd** [cvc5/cvc5#8004] [Q&A] [answered]
Hello,
i use the Java bindings of CVC5 in a [project](https://github.com/sosy-lab/java-smt) and noticed that there seems to be no method to translate a Term from one Solver instance to another (or i missed it in the Java API). Would it be possible to add this feature to the CVC5 bindings?

Thank you for your work!
Best,
Daniel

**Comment by mpreiner** [comment, marked as answer]
Hi @baierd,
Sharing terms between solver instances is currently not supported. However, we are working on a solution to make terms solver-agnostic so that users can reuse terms in different solver instances.

**Comment by baierd** [comment]
Thats nice to hear. Thank you!

**Comment by kfriedberger** [comment]
Is there any update on this topic yet? We would really like to use term-sharing in [JavaSMT](https://github.com/sosy-lab/java-smt).

**Comment by ajreynol** [comment]
Yes, after a recent PR (https://github.com/cvc5/cvc5/pull/9020), you should be able to freely exchange Terms, Sorts, etc. between Solver instances.

Let us know if you run into any issues.

FYI @aniemetz

**Reply by 4tXJ7f** [reply]
With our Java bindings, this is not quite the case yet. We still have a separate `Solver` reference in Terms, etc. there and there will potentially be issues with garbage collection: if terms are created in one solver, used in another, and then the first instance is destroyed. Also, Terms, etc. cannot be shared between solvers created in different threads. This is a limitation that is currently shared between all APIs.

**Comment by kfriedberger** [comment]
Our use-case for CVC5 is as follows:
- We have one special Solver instance "X" for creating and manipulating terms. To keep this description simple: This instance is alive from program start to program termination. We expect all terms to be valid as long as the instance "X" is not yet closed.
- We have several other Solver instances where we push/pop assertions, query satisfiability, unsat cores, and models. We create and close those Solver instances in arbitrary order and arbitrary number. All terms can be used in all alive Solver instances without any explicitly required conversion, until the special Solver instance "X" from above is destroyed.
- Multi-threaded usage would be nice, but is not (yet) used or needed. All Solver instances are currently used in one thread.
- Until now, garbage collection was not yet an issue. Keeping a larger set of terms in memory should not require massive amounts of memory, i.e., it is sufficient if the memory footprint scales with the formula size and formula number.

I assume that CVC5 supports this use-case already. A first small test for CVC5 in JavaSMT was successful. Larger and more detailed test runs will be done soon.

**Comment by kfriedberger** [comment]
Well, I found a case where this situation is problematic for us:
We get terms via model evaluation from solvers, and those terms are no longer valid, when the corresponding solver is closed.
We need to find a way to keep model evaluations for longer than the corresponding solver instance, e.g., somehow copy them over to the special central solver instance.

**Comment by mudathirmahgoub** [comment]
Hi @kfriedberger 
Sharing constant terms across multiple solvers is now supported in the latest version. 
https://github.com/cvc5/cvc5/blob/main/test/unit/api/java/SolverTest.java#L3048
Sharing more complex terms like declare oracles are not supported yet. 
Does this address your concern?

**Reply by kfriedberger** [reply]
At a  first look, this is a good step in the wanted direction.
We actually want to share any kind of term, without the user knowing about it, e.g., this also includes constant symbols and uninterpreted functions and any combination of operators like conjunction, disjunction, negation, addition, equality, etc. . I am not sure, what a "complex term" is in your definition.
I will make some tests to see if the current state is sufficient.

**Reply by mudathirmahgoub** [reply]
An example of a complex term, is a one created by declareOracleFun like this line:
https://github.com/cvc5/cvc5/blob/main/test/unit/api/java/SolverTest.java#L2964
Using this particular term with a solver different than the one used in the body would probably fail with a segmentation fault if the the latter is no longer in memory.

**Reply by mudathirmahgoub** [reply]
@kfriedberger Most of terms supported by SMTLib should work with different solvers. Some terms related to experimental features (like oracles) may not work. 
So the kind of terms you mentioned should work fine. 
If you encounter any issue, let us know.
